### PR TITLE
upstream: Add an addressList() method to Upstream::HostDescription

### PR DESCRIPTION
**Pull request by RyanTheOptimist**
upstream: Add an addressList() method to Upstream::HostDescription so that connection pools (or the Grid) can implement HappyEyeballs.

Future PRs worked on independently will
* Populate this list based on host resolution
* Consume this list in the connection pools

Risk Level: Low
Testing: N/A - Unused currently
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by RyanTheOptimist** [comment]
@mattklein123 we talked about this briefly a while back. Hopefully this is what you were thinking, but please let me know if there's a better way to do this.

David is planning to do the DNS resolution work and I'm planning to do the connection pool work. This PR allows us to meet in the middle without needing to worry about who gets done first.

File: envoy/upstream/host_description.h [hunk 1]
```diff
@@ -119,6 +120,11 @@ class HostDescription {
    */
   virtual Network::Address::InstanceConstSharedPtr address() const PURE;
 
+  /**
+   * @return the address used to connect to the host.
```
Comment (mattklein123) [inline envoy/upstream/host_description.h:124]: Different comment? Maybe a bit about how this is intended to be used?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Yeah this makes sense to me with a small comment. Can you also add a small test for the new method?

/wait

File: envoy/upstream/host_description.h (hunk 1 above)
Comment (RyanTheOptimist) [reply envoy/upstream/host_description.h]: Done.

**Review by RyanTheOptimist — COMMENTED** [review]
> Yeah this makes sense to me with a small comment. Can you also add a small test for the new method?

Thanks! Done.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**